src/openzeppelin/utils/presets/UniversalDeployer.cairo

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts for Cairo v0.4.0b (utils/presets/UniversalDeployer.cairo)
+
+%lang starknet
+
+from starkware.starknet.common.syscalls import get_caller_address, deploy
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.cairo.common.hash import hash2
+from starkware.cairo.common.bool import FALSE, TRUE
+
+@event
+func ContractDeployed(
+    address: felt,
+    deployer: felt,
+    classHash: felt,
+    salt: felt
+) {
+}
+
+@external
+func deployContract{
+    syscall_ptr: felt*,
+    pedersen_ptr: HashBuiltin*,
+    range_check_ptr
+}(
+    classHash: felt,
+    salt: felt,
+    unique: felt,
+    calldata_len: felt,
+    calldata: felt*
+) -> (address: felt) {
+    let (deployer) = get_caller_address();
+
+    tempvar prefix;
+    if (unique == TRUE) {
+        prefix = deployer;
+    } else {
+        prefix = 'UniversalDeployerContract';
+    }
+
+    let (_salt) = hash2{hash_ptr=pedersen_ptr}(prefix, salt);
+
+    let (address) = deploy(
+        class_hash=classHash,
+        contract_address_salt=_salt,
+        constructor_calldata_size=calldata_len,
+        constructor_calldata=calldata,
+        deploy_from_zero=FALSE,
```

Comment:
Shouldn't this be TRUE to make it easy to deploy contracts to the same address on different networks?

Reply:
The way we aimed for replayability across network was making sure the UDC is deployed with `deploy_from_zero=TRUE` guaranteeing the same UDC address everywhere, enabling the properties you mention.

I guess they're almost the same, the only difference I can think if is that if we set it to `TRUE`, then it could collide with another deployer contract given the same salt, class_hash and calldata.

Reply:
Makes sense. The only downside with the current approach is that the UDC does not allow the deployment of an account at the same address as if the account was deploying through the new deploy account transaction coming with 0.10.1.

Reply:
The more I think about this the more I'm convinced that the UDC should cover all use cases if we want it to be broadly adopted and _part of the protocol_. I.e. no one should have to deploy their own deployer for a custom use case.

I see 3 potential use cases for deployment:

1. deploying a contract at an address that cannot be squatted (unique to the deployer) -> this is currently handled by setting `unique=TRUE`
2. deploying a contract at an address that is independent of the deployer but unique for this instance of the UDC -> this is currently handled by setting `unique=FALSE`
3. deploying a contract at an address that is independent of the deployer and the UDC instance (i.e. solely controlled by the class hash, salt and constructor arguments) -> this is currently not supported.

The third use case is what we would need to support deploying accounts at the same address as if they were deployed with the `deploy account transaction`.

While I see a strong need for 1 and 3, I'm less convinced 2 is needed. If you don't care about the address being squatted (i.e. unique) then you don't really care which deployer contract was used.

Based on the above I would suggest to either
- set `deploy_from_zero=TRUE` when `unique=FALSE` which would support 3 but no longer 2
- or find a way to support both 2 and 3 in the UDC.

@martriay what do you think?

Reply:
I agree with your point.

A way to support the three options is to have three entry points - one for each use case.
1. Unique for the deployer address, which will be the current implementation for `unique=TRUE`.
2. Unique space for the UDC, which will be the current implementation for `unique=FALSE`.
3. `deploy_from_zero=TRU` space address, which will be as you said - uses the `deploy_from_zero=TRUE` in the syscall.

Reply:
I think I agree with Julien in that we don't need to support the 2nd use case, changing `deploy_from_zero` to `TRUE` might be the simplest choice.